src/XR/webXRSessionManager.ts

```diff
@@ -340,11 +436,14 @@ export class WebXRSessionManager implements IDisposable {
     /**
      * Updates the render state of the session
      * @param state state to set
+     * @param xrLayer an optional WebXRLayerWrapper that is the session's main xr layer
      * @returns a promise that resolves once the render state has been updated
      */
-    public updateRenderStateAsync(state: XRRenderState) {
-        if (state.baseLayer) {
-            this._baseLayer = state.baseLayer;
+    public updateRenderStateAsync(state: XRRenderStateInit, xrLayer?: Nullable<WebXRLayerWrapper>) {

```

Comment:
First, why is that async? the XRSessions's updateRenderState is not async, and actually returns nothing (unless I missed something?)

Thinking about the future - at certain point we will add more layers, and they will need to be added to the layers array. The current state is not persistent (i.e. not a member of this class), so there will be a need to keep a record of all layers that should be added when calling this function. Should we keep the state somewhere and allow its augmentation?

Also - thinking about the future calls to this function. If I just want to add a new layer, I will need to only update the layers array. So the call would be something like this:

```javascript
sessionManager.updateRenderState({layers: [firstLayer, secondLayer, baseLayer]});
```

Should there be an easier way to "add layer"?

Reply:
Honestly I have no idea why it was made async, looks like when XRSession.updateRenderState was put in webxr.d.ts, [it was defined as returning a Promise<void>](https://github.com/BabylonJS/Babylon.js/commit/f7bb283a47b3d826acb1d8b5e02e3fd0e9b2c788#diff-dae5bb31818a5792dbbff7a6847dd7ee7aa2624abf3547c2f1303dcc4d4027dbR11). I think that may have come from an assumption following the spec which says the method "schedules changes to be applied to the active render state." The spec states now though that the [method is void](https://developer.mozilla.org/en-US/docs/Web/API/XRSession/updateRenderState#return_value), so we should probably change the method to WebXRSessionManager.updateRenderState to reflect that.

I'm realizing I didn't actually pay too much attention to what this method is actually supposed to do. I'll look into how this is supposed to work and allow for added layers, likely through the WebXRFeature you recommended in your other comment.